docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,384 @@
+    match value {
+        true => led.toggle().unwrap(),
+        false => info!("We got something else")
+    }
+}
+```
+
+:::info
+The reason we need all of this is because Rust doesn't allow us to mutably borrow more than once. To use a peripheral (say PWM) inside multiple tasks, we would need to either move it inside the task entirely, or use a mutable reference to it. If we have multiple tasks, though, once we move our peripheral variable *inside* the first task, we can't pass it to another task, because the value was *moved* inside that task completely. And if we wanted to pass it as a mutable reference instead, we would quickly realize that Rust doesn't allow multiple mutable references at once, to avoid concurrent modifications. So this is why we need to either declare a global, static `Mutex` that any task can access, to ensure that the value cannot be modified concurrently by two different tasks, or use channels and keep the peripheral inside the `main` function.
+:::
+
+## Exercises (not final version)
+
+1. Connect an LED to pin GP2 and a photo-resistor to ADC0. Use [KiCad](https://www.kicad.org/) to draw the schematics. (**1p**)
+2. Make the led light up at 25% intensity. (**1p**)
+3. Read the value of the photo-resistor and print it to the console. (**2p**)
+
+TODO: how to see serial messages on computer
+
+:::info
+To be able to print messages to the console, we need to send messages over a simulated serial port to the computer. For this, we will use the USB driver provided by Embassy.
+
+```rust
+use embassy_rp::usb::{Driver, InterruptHandler};
+use embassy_rp::{bind_interrupts, peripherals::USB};
+use log::info;
+
+// Use for the serial over USB driver
+bind_interrupts!(struct Irqs {
+    USBCTRL_IRQ => InterruptHandler<USB>;
+});
+
+
+// The task used by the serial port driver 
+// over USB
+#[embassy_executor::task]
+async fn logger_task(driver: Driver<'static, USB>) {
+    embassy_usb_logger::run!(1024, log::LevelFilter::Info, driver);
+}
+
+#[embassy_executor::main]
+async fn main(spawner: Spawner) {
+    let peripherals = embassy_rp::init(Default::default());
+
+    // Start the serial port over USB driver
+    let driver = Driver::new(peripherals.USB, Irqs);
+    spawner.spawn(logger_task(driver)).unwrap();
+
+    // ...
+
+    info!("message");
+}
+
+```
+:::
+
+:::warning
+Notice that the USB driver also uses an `InterruptHandler` import that could be confused with the `InterruptHandler` used by ADC. Make sure to use different naming conventions for each one, as described in the warning [here](#adc-in-embassy-rs). 
+:::
+4. Depending on the value read from the photo-resistor, brighten or dim the led. The led should shine brighter when there is *less* light in the room. (**2p**)
+5. Remove the photoresistor from the circuit. Instead, put 2 buttons, button1 on GP0 and button2 on GP1. Make it so that when button1 is pressed, the led is dimmed, and when button2 is pressed, the led is brightened. (**3p**)
```

Comment:
I would use it with only one button and move this to the next lab.

Reply:
Used only one button, and modified the exercise to use RGB.